### ISSUE TITLE
Option to change default address space mapping?

### ISSUE THREAD

**Issue by illuhad** [KhronosGroup/SPIRV-LLVM-Translator#1699]
Hi!

For my use case, I am working with multiple toolchains, all of which have the default address space 0 map to the generic address space (e.g. clang HIP or CUDA toolchains), except for SPIR-V.
It would greatly simplify the architecture of my code if this were also true for the LLVM IR representation used by the SPIR-V translator. Without it, I have to manually rewrite the entire IR to change address space 0 to 4 just for the SPIRV-LLVM-Translator.
It seems that the address space mapping that SPIR-V translator uses can be easily modified just by changing the order of the enum values here: https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/ea3ddc186d1e30ade5581c5c00a6b5ef9857655d/lib/SPIRV/SPIRVInternal.h#L181

Would you accept a PR that introduces an option for advanced users to change the address space mapping for such use cases? Or is there something I am missing?

**Comment by MrSidims** [comment]
Hi, wouldn't infer address space pass in llvm added in the tool chain after the translator work for you?

**Comment by MrSidims** [comment]
If not: The map you are mentioning follows the map in SPIR.h from clang. Conceptually it does look like an OK change to add support for different mapping, for example for the one from NVPTX.h , but may be it's more than just about an option. I don't really have strong objections for such change, though I'd like to hear opinion from @svenvh and @bader .

**Comment by illuhad** [comment]
Thanks for the feedback so far!

My interest is mainly the LLVM -> SPIR-V route. I am already using the InferAddressSpaces pass as an optimization. My understanding is however that it does not solve the fundamental issue that my LLVM IR uses 0 as generic address space, because there are cases where this pass cannot infer the address space. So I still have to put in a lot of effort to rewrite the IR. Is this not correct?

**Comment by bader** [comment]
Can't you configure InferAddressSpaces pass to work with a different generic address space mapping?
See https://github.com/intel/llvm/pull/7418 for inspiration.

**Comment by illuhad** [comment]
@bader I don't think it does what I need it to do. My understanding is that the InferAddressSpaces pass tries to minimize the usage of the generic address space (which can be set as a parameter) by trying to infer address spaces that generic pointer uses actually refer to.

My compiler emits LLVM IR using 0 as generic address space instead of 4, as in the SPIR convention. So, what I need is not trying to minimize the usage of one particular address space, I'd need a pass that does a hard exchange of address spaces and change every use of AS 0 to AS 4.
I don't think I can really on the InferAddressSpaces pass to remove *all* generic usage, since there can be cases where the address space cannot be inferred.

Please let me know if I am missing something :-)

My code works perfectly with this very simple path for llvm-spirv translator: https://github.com/illuhad/hipSYCL/blob/feature/generic-llvm-sscp/src/compiler/llvm-to-backend/spirv/llvm-spirv.patch

**Comment by Naghasan** [comment]
> It would greatly simplify the architecture of my code if this were also true for the LLVM IR representation used by the SPIR-V translator

As this is related to your single pass compiler I guess using a different address space map in clang isn't really an option.

> I don't think I can rely on the InferAddressSpaces pass to remove all generic usage, since there can be cases where the address space cannot be inferred.

You can't indeed, no matter what it being done there will cases where generic is required, `InferAddressSpaces` is an optimization pass (and it bails out very quickly).

**Comment by illuhad** [comment]
>  As this is related to your single pass compiler I guess using a different address space map in clang isn't really an option.

That's right. Because the phase where LLVM IR of kernels is extracted piggybacks on the normal clang host compilation, it is difficult to diverge from what clang does by default. This is the main reason why I end up with a lot of AS 0 usage.
Some corrections are of course necessary (e.g.  how arguments are passed into kernels), but those corrections are fairly small compared to the effort of rewriting the entire IR to exchange AS 0 and AS 4.

> You can't indeed, no matter what it being done there will cases where generic is required, InferAddressSpaces is an optimization pass (and it bails out very quickly).

Thanks for the confirmation :)
